DEVINFO.md

```diff
@@ -0,0 +1,187 @@
+
+* **All network input reception is time-bound.**  Every piece of code that ingests data _from the network_ must impose a maximum amount of time that ingestion can take.  If the data takes too long to arrive, then it must be discarded without any further processing.  There is no time bound for data ingested from disk or passed as an argument; this requirement is meant by the space-bound requirement.
+
+* **Untrusted data ingestion must not panic.**  Every piece of code that ingests untrusted data must gracefully handle errors.  Panicking failures are forbidden for such data.  Panics are only allowed if the ingested data was previously written by the node (and thus trusted).
+
+### Non-consensus Changes to Blocks, Microblocks, Transactions, and Clarity
+
+Any changes to code that alters how a block, microblock, or transaction is processed by the node should be **treated as a breaking change until proven otherwise**.  This includes changes to the Clarity VM.  The reviewer _must_ flag any such changes in the PR, and the submitter _must_ convince _all_ reviewers that they will _not_ break consensus.
+
+Changes that touch any of these four code paths must be treated with the utmost care.  If _any_ core developer suspects that a given PR would break consensus, then they _must_ act to prevent the PR from merging.
+
+### Changes to the Peer Network
+
+Any changes to the peer networking code **must be run in production before the PR can be merged.**  The submitter should set up a testable node or set of nodes that reviewers can interact with.
+
+Changes to the peer network should be deployed incrementally and tested by multiple ecosystem entities when possible to verify that they function properly in a production setting.
+
+### Performance Improvements
+
+Any PRs that claim to improve performance **must ship with reproducible benchmarks** that accurately measure the improvement.  This data must also be reported in the PR submission.
+
+### Error Handling
+
+* **Each subsystem must have its own `Error` type.**  Error types of aggregate subsystems are encouraged to both wrap their constituent subsystems' `Error` types in their own `Error` types, as well as provide conversions from them via a `From` trait implementation.
+
+* Functions that act on externally-submitted data **must never panic**.  This includes code that acts on incoming network messages, blockchain data, and burnchain (Bitcoin) data.
+
+* **Runtime panics should be used sparingly**.  Generally speaking, a runtime panic is only appropriate if there is no reasonable way to recover from the error condition.  For example, this includes (but is not limited to) disk I/O errors, database corruption, and unreachable code.
```

Comment:
It's generally okay or even preferable to panic if you run into unexpected conditions during startup (CLI args, config file, socket in use, etc.), right?

Reply:
It's somewhat possible to recover from these things (wait and retry on a socket, use defaults if config is bad). While working a recent PR, I briefly considered loading default values instead of panicking if config file input was out-of-range, so that's what made me think of this

Reply:
> It's somewhat possible to recover from these things (wait and retry on a socket, use defaults if config is bad). 

Disagree strongly.  We do not know the user's intent, nor can we divine it from garbage input.  If we can't act on the input, then the only acceptable thing to do is crash with an informative hint.